File: files/jvm/src/main/java/com/swoval/files/Observers.java
package com.swoval.files;

import com.swoval.files.FileTreeViews.Observer;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Container class that wraps multiple {@link FileTreeViews.Observer} and runs the callbacks for
 * each whenever the {@link PathWatcher} detects an event.
 *
 * @param <T> the data type for the {@link PathWatcher} to which the observers correspond
 */
class Observers<T> implements FileTreeViews.Observer<T>, AutoCloseable {
  private final AtomicInteger counter = new AtomicInteger(0);
  private final Map<Integer, FileTreeViews.Observer<T>> observers = new LinkedHashMap<>();

  @Override
  public void onNext(final T t) {
    final List<FileTreeViews.Observer<T>> cbs;
    synchronized (observers) {
      cbs = new ArrayList<>(observers.values());
    }
    final Iterator<FileTreeViews.Observer<T>> it = cbs.iterator();
    while (it.hasNext()) {
      try {
        it.next().onNext(t);
      } catch (final Exception e) {
        e.printStackTrace();
      }
    }
  }

  @Override
  public void onError(final Throwable throwable) {
    final List<FileTreeViews.Observer<T>> cbs;
    synchronized (observers) {
      cbs = new ArrayList<>(observers.values());
    }
    final Iterator<FileTreeViews.Observer<T>> it = cbs.iterator();
    while (it.hasNext()) {
      try {
        it.next().onError(throwable);
      } catch (final Exception e) {
        e.printStackTrace();
      }
    }
  }

  /**
   * Add an cacheObserver to receive events.
   *
   * @param observer the new cacheObserver
   * @return a handle to the added cacheObserver that can be used to halt observation using {@link
   *     com.swoval.files.Observers#removeObserver(int)} .
   */
  int addObserver(final Observer<T> observer) {
    final int key = counter.getAndIncrement();
    synchronized (observers) {
      observers.put(key, observer);
    }
    return key;
  }

  /**
   * Remove an instance of {@link FileTreeViews.CacheObserver} that was previously added using
   * {@link com.swoval.files.Observers#addObserver(FileTreeViews.Observer)}.
   *
   * @param handle the handle to remove
   */
  void removeObserver(int handle) {
    synchronized (observers) {
      observers.remove(handle);
    }
  }

  @Override
  public void close() {
    observers.clear();
  }
}

File: files/js/src/main/scala/com/swoval/files/Observers.scala
package com.swoval.files

import com.swoval.files.Directory.Entry
import com.swoval.files.Directory.Observer
import com.swoval.files.Directory.OnChange
import com.swoval.files.Directory.OnUpdate
import java.util.Collection
import java.util.HashMap
import java.util.Iterator
import java.util.Map
import java.util.concurrent.atomic.AtomicInteger
import Observers._

private[files] object Observers {

  def apply[T](onchange: OnChange[T]): Observer[T] = new Observer[T]() {
    override def onCreate(newEntry: Entry[T]): Unit = {
      onchange.apply(newEntry)
    }

    override def onDelete(oldEntry: Entry[T]): Unit = {
      onchange.apply(oldEntry)
    }

    override def onUpdate(oldEntry: Entry[T], newEntry: Entry[T]): Unit = {
      onchange.apply(newEntry)
    }
  }

  def apply[T](oncreate: OnChange[T], onupdate: OnUpdate[T], ondelete: OnChange[T]): Observer[T] =
    new Observer[T]() {
      override def onCreate(newEntry: Entry[T]): Unit = {
        oncreate.apply(newEntry)
      }

      override def onDelete(oldEntry: Entry[T]): Unit = {
        ondelete.apply(oldEntry)
      }

      override def onUpdate(oldEntry: Entry[T], newEntry: Entry[T]): Unit = {
        onupdate.apply(oldEntry, newEntry)
      }
    }

}

/**
 * Container class that wraps multiple [[Observer]] and runs the callbacks for each whenever the
 * [[FileCache]] detects an event.
 * @tparam T The data type for the [[FileCache]] to which the observers correspond
 */
private[files] class Observers[T] extends Observer[T] with AutoCloseable {

  private val counter: AtomicInteger = new AtomicInteger(0)

  private val lock: AnyRef = new AnyRef()

  private val observers: Map[Integer, Observer[T]] = new HashMap()

  override def onCreate(newEntry: Entry[T]): Unit = {
    var cbs: Collection[Observer[T]] = null
    lock.synchronized {
      cbs = observers.values
    }
    val it: Iterator[Observer[T]] = cbs.iterator()
    while (it.hasNext) it.next().onCreate(newEntry)
  }

  override def onDelete(oldEntry: Entry[T]): Unit = {
    var cbs: Collection[Observer[T]] = null
    lock.synchronized {
      cbs = observers.values
    }
    val it: Iterator[Observer[T]] = cbs.iterator()
    while (it.hasNext) it.next().onDelete(oldEntry)
  }

  override def onUpdate(oldEntry: Entry[T], newEntry: Entry[T]): Unit = {
    var cbs: Collection[Observer[T]] = null
    lock.synchronized {
      cbs = observers.values
    }
    val it: Iterator[Observer[T]] = cbs.iterator()
    while (it.hasNext) it.next().onUpdate(oldEntry, newEntry)
  }

  def addObserver(observer: Observer[T]): Int = {
    val key: Int = counter.getAndIncrement
    lock.synchronized {
      observers.put(key, observer)
    }
    key
  }

  def removeObserver(handle: Int): Unit = {
    lock.synchronized {
      observers.remove(handle)
    }
  }

  override def close(): Unit = {
    observers.clear()
  }

}
